text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+for no benefit.
+
+For example, imagine this component that checks whether the `files` property is
+supported in input elements:
+
+```js
+import Component from '@ember/component';
+import { computed } from '@ember/object';
+
+export default Component.extend({
+  inputElement: computed(function() {
+    return document.createElement('input');
+  }),
+
+  supportsFiles: computed('inputElement', function() {
+    return 'files' in this.inputElement;
+  }),
+
+  didInsertElement() {
+    if (this.supportsFiles) {
+      // do something
+    } else {
+      // do something else
+    }
+  },
+});
+```
+
+This component would create and retain an `HTMLInputElement` DOM node for the
+lifetime of the component, even though all we really want to cache is the
+Boolean value of whether the browser supports the `files` attribute.
+
+Particularly on mobile devices, where RAM is limited and often slow, we should
+be more conservative about our memory consumption. Tracked properties switch
+from an opt-out caching model to opt-in, allowing developers to err on the side
+of reduced memory usage, but easily enabling caching (a.k.a. memoization) if a
+property shows up as a bottleneck during profiling.
+
+## Detailed Design
+
+This RFC proposes adding the `tracked` decorator function, used to mark
+properties and getters as tracked:
+
+```ts
+const tracked: PropertyDecorator;
+```
+
+This new function will be exported from `@glimmer/tracking`. Revisiting our
+example from earlier, `@tracked` can be used on native class fields and
+getters/setters:
+
+```js
+import { tracked } from '@glimmer/tracking';
+
+class Person {
+  @tracked firstName = 'Tom';
+  @tracked lastName = 'Dale';
+
+  @tracked
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+```
+
+### Getting Tracked Properties
+
+Tracked properties can be accessed using standard Javascript syntax. From the
+user's point of view, there is nothing special about them. This should continue
+to work in the future, even if new methods are added for accessing properties,
+because tracked properties use native getters under the hood.
+
+```js
+let person = new Person();
+
+// Dot notation
+const fullName = person.fullName;
+// Destructuring
+const { fullName } = person;
+// Bracket notation for computed property names
+const fullName = person['fullName'];
+```
+
+### Setting Tracked Properties
+
+Tracked properties can be set using standard Javascript syntax. They use native
+setters under the hood, meaning that there is no need for using a setter method
+like `set`.
+
+```js
+let person = new Person();
+
+// Simple assignment
+person.firstName = 'Jen';
+// Addition assignment (+=)
+person.lastName += 'Weber';
+// Increment operator
+person.age++;
+```
+
+### Autotracking
+
+Tracked properties do not need to specify their dependencies. Under the hood,
+this works by utilizing an _autotrack stack_. This stack is a bit of global
```

Comment:
where does this global state live? or is 'global state' a metaphore?

Reply:
It's local to module scope.